---
title: "Motif distribution on S. pombe genome"
output:
  html_document:
    css: style.css
    fig_height: 6
    fig_width: 12
runtime: shiny
---

```{r library, echo=FALSE, message=FALSE}
options(width = 6000)
library(stringr)
library(pipeR)
library(plyr)
library(dplyr)
library(ggplot2)
library(Biostrings)
library(GenomicRanges)
```

```{r static, echo=FALSE}
revcomp = function(x)
    x %>>% DNAStringSet() %>>% reverseComplement() %>>% as.character()

complement = function(x)
    chartr('ATGC', 'TACG', x)

make_windows = function(len, width, step)
    successiveIRanges(rep(width, (len - width) / step + 1), step - width)

load('data/chromosomes.rda')
```

### パラメータ

```{r input, echo=FALSE}
inputPanel(
    textInput('query', 'Query motif:', 'CGTGCA'),
    sliderInput('width', 'Window width:',
                5000, 50000, 20000, 5000),
    sliderInput('step', 'Sliding step:',
                500, 5000, 2000, 500),
    sliderInput('start', 'View position:',
                0, 5500000, 0, 500000),
    sliderInput('range', 'View width:',
                500000, 5600000, 5600000, 500000)
)
#input = data_frame(query='CGTGCA', width=20000, step=2000)
```

```{r data, echo=FALSE}

make_query = function(s) {
    rc = revcomp(s)
    if (s == rc) {
        PDict(s)
    } else {
        PDict(c(s, rc))
    }
}

.sliding_window = function(.query, .width, .step) {
    .data = ldply(chromosomes, function(.s) {
        .windows = make_windows(length(.s), .width, .step)
        .nucl_count = Views(.s, .windows) %>>%
            alphabetFrequency(as.prob=TRUE) %>>%
            data.frame() %>>% tbl_df() %>>%# (? .) %>>%
            select(A, C, G, T, N) %>>%
            mutate(AT=A + T, GC=C + G)
        .hit = matchPDict(.query, .s) %>>% Biostrings::unlist()
        data.frame(
            pos=end(.windows) - .width / 2,
            .nucl_count,
            observed=countOverlaps(.windows, .hit)
          )
    }, .id='chr') %>>% tbl_df()# %>>% (? .)

    .prob = .query@dict0 %>>%
        as.character() %>>%
        strsplit('') %>>%
        llply(paste, collapse='*') %>>%
        unlist() %>>%
        paste(collapse='+')

    .data = .data %>>%
        mutate_(p = .prob) %>>%
        mutate(expected = p * (.width - width(.query)[1] + 1))
}

.observe = function(.query) {
    ldply(chromosomes, function(.s) {
        matchPDict(.query, .s) %>>%
            Biostrings::unlist() %>>%
            start() %>>%
            (data.frame(pos=.))
    }, .id='chr') %>>% tbl_df()# %>>% (? .)
}

.reactive = reactive({

.query = make_query(input$query)
.data = .sliding_window(.query, input$width, input$step)
.observed = .observe(.query)

.end = min(input$start + input$range, 5.6e6)
.data = .data %>>% filter(pos > input$start, pos < .end)
.observed = .observed %>>% filter(pos > input$start, pos < .end)
.breaks = pretty(c(input$start, .end), 20)

p1 = ggplot(.data, aes(pos))+
    geom_area(aes(y=observed), fill='#009999')+
    geom_line(aes(y=expected), colour='#FF3300', alpha=0.6)+
    geom_point(data=.observed, aes(y=-1), shape='|')+
    facet_grid(chr ~ ., labeller=label_both)+
    scale_x_continuous(expand=c(0, 0), breaks=.breaks)+#seq(0, 6e6, 2e5))+
    theme_bw()+
    theme(axis.text.x=element_text(angle=20))+
    labs(x='Position', y='Count')+
    ggtitle(sprintf('%s|%s, width = %d, step = %d',
                    input$query, revcomp(input$query),
                    input$width, input$step))

.lambda = mean(.data$observed)
.mids = seq(0, max(.data$observed))
.breaks = .mids - 0.5

p2= ggplot(.data, aes(observed))+
    geom_histogram(aes(y=..density..), breaks=.breaks, fill='#009999')+
    geom_line(aes(x=x, y=y), colour='#FF3300',
              data=data.frame(x=.mids, y=dpois(.mids, .lambda)))+
    labs(x='motifs per window',
         title=sprintf('poisson(lambda=%.3f)', .lambda))

.intervals = .observed %>>%
    group_by(chr) %>>%
    arrange(pos) %>>%
    do(data_frame(distance=diff(.$pos)))
.breaks = pretty(range(.intervals$distance), 30)
.bin = diff(.breaks)[1]
.x = .breaks + .bin / 2
.y = dexp(.x, 0.25 ^ 6)

p3 = .intervals %>>%
    ggplot(aes(distance))+
    geom_histogram(aes(y=..density..), breaks=.breaks, fill='#009999')+
    geom_line(aes(y=count),
              data=data_frame(distance=.x, count=.y),
              colour='#FF3300')+
    labs(x='distance between motifs',
         title=sprintf('exp(lambda=%.3f)', .lambda))

list(p1=p1, p2=p2, p3=p3)
})
```

<b>
<font color="#009999">green = observed</font>,
<font color="#FF3300">red = expected</font>
</b>

### モチーフの分布

```{r plot, echo=FALSE}
renderPlot(.reactive()$p1, width=3000, height=1200, res=200)

#https://github.com/hadley/ggplot2/pull/939
ggpng = function(..., width, height) {
    grDevices::png(..., width = width, 
        height = height, res = 300, units = "in")
}

downloadHandler(
    filename = function(){
        sprintf('%s-w%d-s%d.png', input$query, input$width, input$step)
    },
    content = function(file) {
        ggsave(file, .reactive()$p1, device=ggpng,
               width=10, height=3, scale=2)
    }
)
```

### ウィンドウあたりのモチーフ数

```{r perwindow, echo=FALSE}
renderPlot(.reactive()$p2, width=2000, height=800, res=200)

downloadHandler(
    filename = function(){
        sprintf('%s-w%d-s%d-perwindow.png', input$query, input$width, input$step)
    },
    content = function(file) {
        ggsave(file, .reactive()$p2, device=ggpng,
               width=10, height=3, scale=2)
    }
)
```

### モチーフ間の距離 (bp)

```{r interval, echo=FALSE}
renderPlot(.reactive()$p3, width=2000, height=800, res=200)

downloadHandler(
    filename = function(){
        sprintf('%s-w%d-s%d-interval.png', input$query, input$width, input$step)
    },
    content = function(file) {
        ggsave(file, .reactive()$p2, device=ggpng,
               width=10, height=3, scale=2)
    }
)
```
