---
title: "Motif occurrence in S. pombe reference genome"
output:
  html_document:
    css: style.css
runtime: shiny
---

Windows are 20000bp wide and not overlapped.

```{r library, echo=FALSE, message=FALSE}
library(stringr)
library(pipeR)
library(plyr)
library(dplyr)
library(ggplot2)
library(Biostrings)
library(GenomicRanges)
library(doMC)
doMC::registerDoMC(min(parallel::detectCores(), 12))
```

```{r static, echo=FALSE}
revcomp = function(x)
    x %>>% DNAStringSet() %>>% reverseComplement() %>>% as.character()

complement = function(x)
    chartr('ATGC', 'TACG', x)

make_windows = function(len, width, step)
    successiveIRanges(rep(width, (len - width) / step + 1), step - width)

make_query = function(s) {
    rc = revcomp(s)
    if (s == rc) {
        PDict(s)
    } else {
        PDict(c(s, rc))
    }
}

.sliding_window = function(.query, .width, .step) {
    .data = ldply(chromosomes, function(.chr) {
        .windows = make_windows(length(.chr), .width, .step)
        .hit = matchPDict(.query, .chr) %>>% Biostrings::unlist()
        data.frame(
            observed=countOverlaps(.windows, .hit)
        )
    }, .id='chr') %>>% tbl_df()# %>>% (? .)
}
```

```{r data, echo=FALSE}
#setwd('~/git/pomber/rmd')
.cache = 'data/coefvar-ref.csv.gz'
if (file.exists(.cache)) {
    .data = read.table(.cache, header=TRUE, stringsAsFactors=FALSE) %>>% tbl_df()
} else {

load('data/ensembl.rda')

.weight = alphabetFrequency(chromosomes) %>>%
    data.frame() %>>%
    select(matches('[ATGCN]')) %>>%
    summarise_each(funs(sum))
.null = paste(sample(names(.weight), 2000000, replace=TRUE, .weight), collapse='')
chromosomes = c(chromosomes, DNAStringSet(c(randomized=.null)))

.weight = .weight %>>% select(matches('[ATGC]'))
.null = paste(sample(names(.weight), 2000000, replace=TRUE, .weight), collapse='')
chromosomes = c(chromosomes, DNAStringSet(c(without_N=.null)))

if (FALSE) {
    .query = 'ATGCGC'
    .data = .sliding_window(make_query(.query), .width, .step)
    .data %>>%
        group_by(chr) %>>%
        summarise(mean=mean(observed), var=var(ovserved), sd=sd(observed), vm_ratio=var/mean, cv=sd/mean)
}

.oligos = strsplit('ATGC', '') %>>%
    rep(6) %>>%
    expand.grid() %>>%
    mlply(function(...) paste0(...)) %>>%
    unlist(use.names=FALSE)
names(.oligos) = .oligos

.queries = data_frame(motif=.oligos, revcomp=revcomp(.oligos)) %>>%
    mutate(motif=pmin(motif, revcomp)) %>>%
    select(-revcomp) %>>%
    filter(!duplicated(motif)) %>>%
    mutate(revcomp=revcomp(motif), motif=ifelse(motif==revcomp, motif, paste(motif, revcomp, sep='|'))) %>>%
    select(-revcomp)

.width = 20000
.step = 20000

.data = mdply(.queries, function(motif) {
    .query = strsplit(motif, '\\|') %>>% unlist() %>>% PDict()
    .sliding_window(.query, .width, .step) %>>%
    group_by(chr) %>>%
    summarise(mean=mean(observed), var=var(observed), sd=sd(observed), vm_ratio=var/mean, cv=sd/mean)
}, .id='motif', .parallel=TRUE)

if (FALSE) {
    .data %>>%
        tbl_df() %>>%
        filter(grepl('I', chr)) %>>%
        arrange(cv) %>%
        each(head, tail)(.)
}

write.table(.data, gzfile(.cache), sep='\t', row.names=FALSE, quote=FALSE)
}
```

```{r input, echo=FALSE}
.chrs = unique(.data$chr)
.chrs = .chrs[.chrs != 'without_N']
names(.chrs) = .chrs
inputPanel(
    checkboxGroupInput('chr', 'Chromosome:', .chrs, .chrs),
    sliderInput('ubound', 'Display x max:', 5, 150, 150, 5)
)
```

```{r render, echo=FALSE}
filtered = reactive({
    subset(.data, chr %in% input$chr) %>>% filter(mean < input$ubound)
})

.range = reactive(range(filtered()$mean))

renderPlot(filtered() %>>%
    ggplot(aes(mean, var))+
    geom_point(aes(colour=chr), alpha=0.6)+
    geom_line(data=data_frame(mean=.range(), var=.range()))+
    theme_bw()
)

if (FALSE) {
    renderPlot(filtered() %>>%
        ggplot(aes(mean, sd))+
        geom_point(aes(colour=chr), alpha=0.6)+
        geom_line(data=data_frame(mean=.range(), sd=.range()))+
        theme_bw()
    )
}

renderDataTable(filtered())
#renderDataTable(filtered(), options=list(order=c(4, 'desc')))
```
