---
title: "Motif distribution on S. pombe genome"
output:
  html_document:
    css: style.css
    fig_height: 6
    fig_width: 12
runtime: shiny
---

```{r library, echo=FALSE, message=FALSE}
options(width = 6000)
library(stringr)
library(pipeR)
library(plyr)
library(dplyr)
library(ggplot2)
library(Biostrings)
library(GenomicRanges)
```

```{r static, echo=FALSE}
revcomp = function(x)
    x %>>% DNAStringSet() %>>% reverseComplement() %>>% as.character()

complement = function(x)
    chartr('ATGC', 'TACG', x)

make_windows = function(len, width, step)
    successiveIRanges(rep(width, (len - width) / step + 1), step - width)

load('data/chromosomes.rda')
```

## パラメータ

```{r input, echo=FALSE}
inputPanel(
    textInput('query', 'Query motif:', 'GCGCAC'),
    sliderInput('width', 'Window width:',
                2000, 50000, 10000, 2000),
    sliderInput('step', 'Sliding step:',
                100, 2000, 1000, 100),
    sliderInput('start', 'View position:',
                0, 5500000, 0, 500000),
    sliderInput('range', 'View width:',
                500000, 5600000, 5600000, 500000)
)
```

```{r data, echo=FALSE}

make_query = function(s) PDict(c(s, revcomp(s)))

.sliding_window = function(.query, .width, .step) {
    .data = ldply(chromosomes, function(.s) {
        .windows = make_windows(length(.s), .width, .step)
        .nucl_count = Views(.s, .windows) %>>%
            alphabetFrequency(as.prob=TRUE) %>>%
            data.frame() %>>% tbl_df() %>>%# (? .) %>>%
            select(A, C, G, T, N) %>>%
            mutate(AT=A + T, GC=C + G)
        .hit = matchPDict(.query, .s) %>>% Biostrings::unlist()
        data.frame(
            pos=end(.windows) - .width / 2,
            .nucl_count,
            observed=countOverlaps(.windows, .hit)
          )
    }, .id='chr') %>>% tbl_df()# %>>% (? .)

    .prob = .query@dict0 %>>%
        as.character() %>>%
        strsplit('') %>>%
        llply(paste, collapse='*') %>>%
        unlist() %>>%
        paste(collapse='+')

    .data = .data %>>%
        mutate_(p = .prob) %>>%
        mutate(expected = p * (.width - width(.query)[1] + 1))
}

.observe = function(.query) {
    ldply(chromosomes, function(.s) {
        matchPDict(.query, .s) %>>%
            Biostrings::unlist() %>>%
            start() %>>%
            (data.frame(pos=.))
    }, .id='chr') %>>% tbl_df()# %>>% (? .)
}

#input = data_frame(query='GCGCAC', width=10000, step=1000)

.reactive = reactive({

.query = make_query(input$query)
.data = .sliding_window(.query, input$width, input$step)
.observed = .observe(.query)

.end = min(input$start + input$range, 5.6e6)
.data = .data %>>% filter(pos > input$start, pos < .end)
.observed = .observed %>>% filter(pos > input$start, pos < .end)
.breaks = pretty(c(input$start, .end), 20)

#.breaks = seq_len(max(.data$observed) + 1) - 1.5
#ggplot(.data, aes(observed))+
#    geom_histogram(breaks=.breaks)

.p1 = ggplot(.data, aes(pos))+
    geom_area(aes(y=observed), fill='#009999')+
    geom_line(aes(y=expected), colour='#FF3300', alpha=0.6)+
    geom_point(data=.observed, aes(y=-0.5), shape='|')+
    facet_grid(chr ~ ., labeller=label_both)+
    scale_x_continuous(expand=c(0, 0), breaks=.breaks)+#seq(0, 6e6, 2e5))+
    theme_bw()+
    theme(axis.text.x=element_text(angle=20))+
    labs(x='Position', y='Count')+
    ggtitle(sprintf('%s|%s, width = %d, step = %d',
                    input$query, revcomp(input$query),
                    input$width, input$step))

.p2 = .observed %>>%
    group_by(chr) %>>%
    arrange(pos) %>>%
    do(data_frame(distance=diff(.$pos))) %>>%
    ggplot(aes(distance))+
    geom_histogram()

list(p1=.p1, p2=.p2)
})
```

## モチーフの分布

green = observed, red = expected

```{r plot, echo=FALSE}
renderPlot(.reactive()$p1, width=3000, height=1200, res=200)

#https://github.com/hadley/ggplot2/pull/939
ggpng = function(..., width, height) {
    grDevices::png(..., width = width, 
        height = height, res = 300, units = "in")
}

downloadHandler(
    filename = function(){
        sprintf('%s-w%d-s%d.png', input$query, input$width, input$step)
    },
    content = function(file) {
        ggsave(file, .reactive()$p1, device=ggpng,
               width=10, height=3, scale=2)
    }
)
```

## 発見されたモチーフ間の距離

```{r hist, echo=FALSE}
renderPlot(.reactive()$p2, width=2000, height=800, res=200)

downloadHandler(
    filename = function(){
        sprintf('%s-w%d-s%d-hist.png', input$query, input$width, input$step)
    },
    content = function(file) {
        ggsave(file, .reactive()$p2, device=ggpng,
               width=10, height=3, scale=2)
    }
)
```
